### ISSUE TITLE
The book takes too long to typecheck

### ISSUE THREAD

**Issue by mrakgr** [martinescardo/HoTT-UF-Agda-Lecture-Notes#18]
It might be good to split it into multiple modules, or to make it into a library with the exercises placed separately because right now it really does take too long. It would also be good design for the exercises to not have the reader backtrack to them after the later chapters, and instead introduce them after relevant concepts are introduced such as for the `'x ≤ y' if and only if 'Σ \(z : ℕ) → x + z ≡ y'` and `(x ≤ y) ≡ Σ \(z : ℕ) → x + z ≡ y` exercises.

Also I had an issue with figuring out what fonts to use yesterday. Right now I am using DejaVu Sans Mono as the default and Symbola as [the fallback](https://stackoverflow.com/questions/40672995/how-to-fix-the-bool-symbol-from-agda-showing-up-as-a-square-box-with-hex-and-dec). I am wondering what fonts you are using?

**Comment by martinescardo** [comment]
Yes, it takes 34s in a 6th generation core i5 laptop. Sorry about that. But you should need to typecheck it only once. If you want to do something building on that, you can simply import the file, which shouldn't take long if it is already type checked.

Originally I had a number of modules (as you can see from the git history). However, because I want to also generate a pdf file with *internal* links, I decided to move to a monollithic development (search for monolithic in the git history).

The process to generate the pdf (given that we also want html) is this:

  * Write the file in markdown format, with Agda code in code environments.
  * Use Agda to generate markdown with embedded html from the Agda definitions.
  * Use Jekyll to convert the markdown with html into pure html
  * Use google chrome to convert the html to pdf 

There are many html-to-pdf converters, but chrome seems to be the only one that can convert html internal links to pdf internal links. But this works only within *one* pdf file, and this is why I decided to have only one file. But also one has to distinguish between expositions and libraries. This is intended to be an exposition. In my own development, I have all this material split in several modules, as one would expect. These notes are a book, not a library. In any case, I don't want to be normative. I am just explaining why I did it the way I did it.

**Comment by martinescardo** [comment]
And your comments are appreciated.

**Comment by martinescardo** [comment]
Regarding the exercises, you are welcome to produce pull requests. From the point of view of the writer, there is a balance between keeping the flow of the presentation and the flow of required exercising. In books, we have expressions such as "exercise left to the reader", which are then used later in proofs of theorems. When you try to do this in Agda, this becomes tricky. The way I did it was to in some instances give the definitions of the things to be proved as type declarations, and then much later (in the sample solutions), as definitions. However, for the particular exercise you mention, this is not possible, because the exercise mentions something we know in natural language (existence) which hasn't been defined in Agda yet. In this particular instance, I chose to *not* give a solution in a list of solutions, but instead use the solution as an opportunity to teach (for people who didn't solve the exercise). In any case, all exercises are for teaching, and it is a matter of art to decided whether you include their solution in the main exposition or in separate sections with solutions. I am not claiming I have obtained an optimal solution to this didactic question. But I tried my best,

**Comment by martinescardo** [comment]
Regarding fonts, in .emacs I use
```
(set-fontset-font "fontset-default" nil
                  (font-spec :name "DejaVu Sans"))
```

**Comment by martinescardo** [comment]
As I said, your comments and suggestions are very welcome. I would suggest that next time you formulate three questions/proposals, you do it in three separate issues, unless you feel they are related. You are also welcome and encouraged to argue here if you disagree and/or have a better proposal after my explanations.

**Comment by martinescardo** [comment]
One last remark. You will have noticed that in some solved exercises I give a type declaration of something to be defined, but that in other solved exercises I instead explain the exercise informally (but hopefully rigorously). This is *deliberate*, so that students learn not only how to *prove* theorems in Agda but also how to *formulate* them. One thing I feel guilty about these notes is the lack of *more* exercises. I will work on that (and I have been working on that).

**Comment by martinescardo** [comment]
I should also report something I have learned when trying to transform a maths library into a maths book. In books, we feel free to prove the same theorem many times, not only to illustrate different proofs, but also to get increasing generality in a step-by-step way. It seems that libraries try to first prove the most general version of theorems instead. This makes the structuring of libraries very awkward. I think the book way is good not only for didactical purposes but also in practice, to get a better modularization of mathematics. One should not be afraid of proving the same theorem more than once in different contexts, or in increasing generality rather than in maximum generality.

**Comment by mrakgr** [comment]
> Regarding fonts, in .emacs I use
> 
> ```
> (set-fontset-font "fontset-default" nil
>                   (font-spec :name "DejaVu Sans"))
> ```

This was what I was using before **Edit**: as a fallback, but I had a lot of boxes with that. What are you using as the default font in Emacs? I'll be surprised if you say `DejaVu Sans Mono`.

**Comment by mrakgr** [comment]
> Yes, it takes 34s in a 6th generation core i5 laptop. Sorry about that. But you should need to typecheck it only once. If you want to do something building on that, you can simply import the file, which shouldn't take long if it is already type checked.
> 
> Originally I had a number of modules (as you can see from the git history). However, because I want to also generate a pdf file with _internal_ links, I decided to move to a monollithic development (search for monolithic in the git history).
> 
> The process to generate the pdf (given that we also want html) is this:
> 
> * Write the file in markdown format, with Agda code in code environments.
> * Use Agda to generate markdown with embedded html from the Agda definitions.
> * Use Jekyll to convert the markdown with html into pure html
> * Use google chrome to convert the html to pdf
> 
> There are many html-to-pdf converters, but chrome seems to be the only one that can convert html internal links to pdf internal links. But this works only within _one_ pdf file, and this is why I decided to have only one file. But also one has to distinguish between expositions and libraries. This is intended to be an exposition. In my own development, I have all this material split in several modules, as one would expect. These notes are a book, not a library. In any case, I don't want to be normative. I am just explaining why I did it the way I did it.

I feel that you could have skipped the HTML parts entirely and just stuck to markdown if links are solely what you desired. It is possible to link to places within the same document in markdown using headers and it is easy enough to generate the TOC automatically from them.

The way the book is structured currently could be far better. For example on line 1630 here is how it looks like in the Agda literal file.

```
<blockquote>
    Theorem. Absurdity of absurdity of absurdity is equivalent to absurdity.
</blockquote>
<blockquote>
    Proof. <em>Firstly</em>, since implication of the assertion &#119910; by the
    assertion &#119909; implies implication of absurdity of &#119909; by absurdity
    of &#119910;, the implication of <em>absurdity of absurdity</em> by <em>truth</em>
    (which is an established fact) implies the implication of
    <em>absurdity of truth</em>, that is to say of <em>absurdity</em>, by <em>absurdity
    of absurdity of absurdity</em>. <em>Secondly</em>, since truth of an assertion
    implies absurdity of its absurdity, in particular truth of
    absurdity implies absurdity of absurdity of absurdity.
</blockquote>
```

I had to switch to the HTML version just to read this as the tags are so distracting.

On the other hand, this on line 911 appears raw in the HTML version.

```
*Exercise.* [Write it]((HoTT-UF-Agda.html#someexercisessol) using
`ℕ-induction`, recursion or iteration, as appropriate.
```

As a template for a good design of an Agda book, I would really recommend looking into [PLFA](https://plfa.github.io/).

If you take a look at how the individual chapters are structured such as [Relations](https://github.com/plfa/plfa.github.io/blob/dev/src/plfa/Relations.lagda) for example, you will see that the Agda literal file has very little apart from actual unicode text and where markdown is used, it is minimally invasive to those going through the book in Emacs. I'd recommend you to focus on getting the literal file to be as clean as possible before trying to deal with the web version as that part is just eye candy and those who are really interested will be spending their time in Emacs and the literal files, so you do not want HTML tags littering them.

The chapters are split into individual modules which aids compilation times, and the exercises can be taken as one goes along inside the same file. After experiencing PLFA I admit I've been spoiled by this, and have been surprised to find a book that does not follow that approach.

If you want the monolithic approach despite this, I really recommend adding some instructions acknowledging the compile times and that the user open a separate file for the exercises. Yesterday that is what I did of course, but only after half an hour of being annoyed by compilation times.

**Comment by mrakgr** [comment]
Adding to that last point - in addition to how the book is intended to be digested, I'd put in some instruction regarding what fonts are to be used and where to find the source. Right now the link to the source is in an awkward place at the end of the TOC. You'll note that PLFA has a link to the source at the top of every chapter and this is also a good practice. The way I went through the PLFA book was to literally get each individual source and paste it in a file. This sort of structure makes it easy for that.

If the reader is intended to find the Github repo and clone it, then it would be good to instruct that. The important thing to remember is that the reader will be starting from the HTML version, but he really wants to be following along in Emacs and the literal file **snip** and the book should be structured as to smooth that transition.

As a side comment, at this point in time getting Agda and its standard library installed along with Emacs has a decent bit of intermediate steps that could be removed. Somehow I did not have trouble getting it to work with Emacs, but I expected it to be a lot worse.

**Comment by martinescardo** [comment]
You are not supposed to read the source code. You are supposed to read this:
https://www.cs.bham.ac.uk/~mhe/HoTT-UF-in-Agda-Lecture-Notes/index.html 
or this 
https://www.cs.bham.ac.uk/~mhe/HoTT-UF-in-Agda-Lecture-Notes/HoTT-UF-Agda.pdf

Agda generates html for the code, and not markdown. See the script "build" to see how this works.

And then there is a version with the agda code extracted, in the folder /agda/ here.

Thanks for your comments.
